RBF Timelines

mononaut opened the pull request:
This PR attempts to better visualize RBF events.

It adds a new section to the transaction page (where relevant), showing a timeline of replacements:

<img width="1254" alt="RBF Transaction Page" src="https://user-images.githubusercontent.com/83316221/207739595-1c141454-a681-45b7-980f-48458287d066.png">

Each 'node' on the timeline is a different version of the transaction, with the one currently being viewed circled in purple. Clicking on a node navigates to that tx version.

If a version has been mined, the corresponding node is circled in green.

The timeline scrolls horizontally to fit longer sequences of replacements, and updates via websocket as subsequent replacements come in.

The PR also adds a new page listing (up to 25 sets of) the latest RBF events, using the same timeline components:

<img width="1254" alt="RBF List" src="https://user-images.githubusercontent.com/83316221/207739646-d813bb23-fa33-49eb-86c1-ee03c3ffa01e.png">

The listing updates in real time via websocket, and includes a toggle to switch between all RBF events and only Full RBF replacements if the instance is running Full RBF.

The page isn't linked anywhere on the main site yet. Perhaps we could add a mini widget version to one of the dashboards?

### Limitations

#### ~Multi-replacements~

~Technically, a single transaction can replace multiple others by fee, which would turn our linear timeline into more like a tree structure. In practice, that doesn't seem to happen very often, if ever.~

~This implementation breaks any trees into separate linear leaf-to-root timelines, and only shows one such timeline on each transaction page.~

_(edit)_ The PR now supports arbitrary trees of replacements, e.g:

<img width="1145" alt="Screenshot 2022-12-16 at 4 56 58 PM" src="https://user-images.githubusercontent.com/83316221/208249921-2f25720c-78b9-41df-8345-ec8841e06e65.png">

#### Pagination

Grouping replacements into timelines makes sorting by recency unstable, so pagination is difficult.

### Todo
 - [x] Support for multi-replacement trees
 - [x] Tooltips
 - [x] Save RBF cache to disk to preserve across restarts
 
### Extensions for future PRs
 - Link to `/rbf` page
 - Pagination / infinite scroll
 - Reverse timeline direction for RTL locales

0xB10C commented:
This is cool - thanks for working on this!

> Multi-replacements
> Technically, a single transaction can replace multiple others by fee, which would turn our linear timeline into more like a tree structure. In practice, that doesn't seem to happen very often, if ever.

I disagree with this. Looking at the RBF data I've collected over the past years it's quite common that e.g. two or even multiple (like 20+ transctions being replaced) transactions are merged into one. Happy to extract some statistics on this if you're interested. IMO a tree like structure is the way to go here.

> Pagination
> Grouping replacements into timelines makes sorting by recency unstable, so pagination is difficult.

Right. Might make sense have two ways to sort here:
- by most recent replacement (unstable pagination)
- by time the first version of this replacement sequence was seen (stable pagination)

mononaut commented:
Thanks for the feedback!

> I disagree with this. Looking at the RBF data I've collected over the past years it's quite common that e.g. two or even multiple (like 20+ transactions) transactions are merged into one. Happy to extract some statistics on this if you're interested. IMO a tree like structure is the way to go here.

Admittedly this was based on only a very unscientific survey of the last few days. I'd love to see some proper statistics, if it's not too much of a hassle to extract.

0xB10C commented:
This are stats based on the replacements my Bitcoin Core node has seen between 2022-04-01 and 2022-11-01. There were a total of 1344596 replacements during these 7 months. My intuition with "quite common" in https://github.com/mempool/mempool/pull/2847#issuecomment-1352999071 might have been biased a bit as  I've spent the most time looking at the more interesting multi-replacements. You'll be fine in >98% of the cases with a simple timeline.


| number of tx replaced by a replacement 	| count   	| share   	|
|----------------------------------------	|---------	|---------	|
| 1                                      	| 1324815 	| 98.52%  	|
| 2                                      	| 12174   	| 0.9054% 	|
| 3                                      	| 6726    	| 0.5002% 	|
| 4                                      	| 586     	| 0.0435% 	|
| 5                                      	| 130     	| 0.0096% 	|
| 6                                      	| 58      	| 0.0043% 	|
| 7                                      	| 20      	| 0.0014% 	|
| 8                                      	| 18      	| 0.0013% 	|
| 9                                      	| 12      	| 0.0008% 	|
| 10                                     	| 8       	|         	|
| 11                                     	| 2       	|         	|
| 12                                     	| 4       	|         	|
| 13                                     	| 4       	|         	|
| 14                                     	| 1       	|         	|
| 15                                     	| 3       	|         	|
| 16                                     	| 4       	|         	|
| 19                                     	| 2       	|         	|
| 21                                     	| 2       	|         	|
| 22                                     	| 1       	|         	|
| 23                                     	| 2       	|         	|
| 24                                     	| 1       	|         	|
| 25                                     	| 10      	|         	|
| 26                                     	| 3       	|         	|
| 29                                     	| 1       	|         	|
| 31                                     	| 1       	|         	|
| 32                                     	| 1       	|         	|
| 39                                     	| 1       	|         	|
| 41                                     	| 1       	|         	|
| 43                                     	| 1       	|         	|
| 54                                     	| 1       	|         	|
| 93                                     	| 1       	|         	|
| 94                                     	| 1       	|         	|
| 97                                     	| 1       	|         	|

mononaut commented:
awesome, thank you!

mononaut commented:
Rebased on master, fixed some bugs, and removed dependency on PR #2845.

softsimon commented:
Suggestions

* Store and restore RBF history in JSON disk cache
* Change the tooltip to typical dual column design
* Display the time stamps like "5 minutes" without the "After" between replacements
* Change "Replacements" title to "RBF History" (?)

<img width="235" alt="Screenshot 2023-03-05 at 13 50 19" src="https://user-images.githubusercontent.com/8561090/222942291-24904488-9977-4b59-866f-e8a51a733fc5.png">

Bugs:

* When new replacement arrives and clicking on the link, the timeline graph highlight doesn't always update.

mononaut commented:
Implemented

- [x] Store and restore RBF history in JSON disk cache
- [x] Change the tooltip to typical dual column design
- [x] Display the time stamps like "5 minutes" without the "After" between replacements
- [x] Change "Replacements" title to "RBF History" (?)
- [x] If you open a transaction with many RBF history points, the current TX point is hidden in view, so always scroll the RBF timeline so the current TX bullet is in view.

Fixed:

- [x] When new replacement arrives and clicking on the link, the timeline graph highlight doesn't always update.
- [x] It looks like the E2E "shows RBF transactions properly" fail and needs to be updated

softsimon commented:
@mononaut please rebase on master

mononaut commented:
Rebased on master, made compatible with recent disk cache changes, and added some other minor frontend improvements.

softsimon commented:
Getting a couple of frontend errors when clicking around between RBF transactions

<img width="1408" alt="Screenshot 2023-03-30 at 11 18 28" src="https://user-images.githubusercontent.com/8561090/228711120-9dc5589e-7b70-4704-94fc-d056c13d1442.png">

softsimon commented on frontend/src/app/components/transaction/transaction.component.html:
```diff
@@ -1,7 +1,7 @@
 
   <div class="title-block">
-    <div *ngIf="rbfTransaction" class="alert alert-mempool" role="alert">
+    <div *ngIf="rbfTransaction && !tx.status.confirmed" class="alert alert-mempool" role="alert">
```
Seems like this require some tx? null check

mononaut replied:
yep, I just pushed a fix

softsimon reviewed (approved):
tested ACK @ [c9ed2e1]

mononaut commented:
rebased on current master [8db7326]
